server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -637,9 +642,34 @@ public void beforeIndexShardRecovery(IndexShard indexShard, IndexSettings indexS
                 }
             }
         };
+        final IndexEventListener afterIndexCreated = new IndexEventListener() {
+            @Override
+            public void afterIndexCreated(IndexService indexService) {
```

Comment:
My first impl had this code inside `afterIndexShardRecovery` event listener, that is a better place IMO, but doing so I incurred in `IllegalStateException: Some shards are still open after the threadpool terminated. Something is leaking index readers or store references.` even if I did call `hasValueReader.close()`. Am I missing something here?